gqlalchemy/vendors/database.py

```diff
@@ -0,0 +1,316 @@
+# Copyright (c) 2016-2022 Memgraph Ltd. [https://memgraph.com]
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Iterator, List, Optional
+
+from ..connection import Connection
+from ..exceptions import GQLAlchemyError
+from ..models import (
+    Constraint,
+    Index,
+    Node,
+    Relationship,
+)
+
+
+class Database(ABC):
+    def __init__(
+        self,
+        host: str,
+        port: int,
+        username: str,
+        password: str,
+        encrypted: bool,
+        client_name: str,
+    ):
+        self._host = host
+        self._port = port
+        self._username = username
+        self._password = password
+        self._encrypted = encrypted
+        self._client_name = client_name
+        self._cached_connection: Optional[Connection] = None
```

Comment:
Could the database class be a good candidate to be a template class of sort? As of now, we should dictate the type of this `_cached_connection` in all of the derived classes, so that we don't end using neo4j connection in memgraph client and vice versa